src/delegate.ts

```diff
@@ -0,0 +1,38 @@
+import { Handle } from 'dojo-core/interfaces';
+import on, { ExtensionEvent } from 'dojo-core/on';
+
+export default function delegate(target: HTMLElement, selector: string, type: string, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: ExtensionEvent, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: (string | ExtensionEvent)[], listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: any, listener: (event: UIEvent) => void): Handle {
+	function matches(target: EventTarget) {
+		let node = <any> target;
+		let matchMethod = (function() {
+			if (typeof node.matches === 'function') {
+				return 'matches';
+			}
+			if (typeof node.webkitMatchesSelector === 'function') {
+				return 'webkitMatchesSelector';
+			}
+			if (typeof node.msMatchesSelector === 'function') {
+				return 'msMatchesSelector';
+			}
+		})();
+
+		while (!node[matchMethod](selector)) {
+			if (!node.parentNode || !node.parentNode[matchMethod]) {
+				return false;
+			}
+			node = node.parentNode;
+		}
+
+		return node;
+	}
+
+	return on(<EventTarget> target, type, function (event: Event) {
+		let matchedEventTarget = matches(event.target);
+		if (matchedEventTarget) {
```

Comment:
Unfortunately, `matches` is subject to the same undesirable behavior as `querySelector` itself, which we probably want to avoid tripping on.

Test case:
1. Create a 3-level element structure, giving the top element a `className` of `ancestor` and the bottom element a `className` of `child`
2. Set up a delegated event handler on the *middle* element, using the selector `.ancestor .child`
3. You would normally expect this *not* to match, given that `ancestor` is only on an element *above* the given root, but it *will* match.

@neonstalwart suggested adding a `contains` check to avoid this (i.e. `if (matchedEventTarget && dom.contains(target, matchedEventTarget))`).

We should also add a test for this case based on the above steps.

(Note: I have to fix a bug in `dom.contains` for this to work in all cases; I should have that fixed sometime this morning.)